group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java

```diff
@@ -966,6 +979,53 @@ private static void maybeUpdateSubscribedTopicNames(
         }
     }
 
+    /**
+     * Updates the subscription count.
+     *
+     * @param oldMember             The old member.
+     * @param newMember             The new member.
+     *
+     * @return Copy of the map of topics to the count of number of subscribers.
+     */
+    public Map<String, Integer> computeSubscribedTopicNames(
+        ConsumerGroupMember oldMember,
+        ConsumerGroupMember newMember
+    ) {
+        Map<String, Integer> subscribedTopicNames = new HashMap<>(this.subscribedTopicNames);
+        maybeUpdateSubscribedTopicNames(
+            subscribedTopicNames,
+            oldMember,
+            newMember
+        );
+        return subscribedTopicNames;
+    }
+
+    /**
+     * Compute the subscription type of the consumer group.
+     *
+     * @param subscribedTopicNames      A map of topic names to the count of members subscribed to each topic.
+     *
+     * @return {@link SubscriptionType#HOMOGENEOUS} if all members are subscribed to exactly the same topics;
+     *         otherwise, {@link SubscriptionType#HETEROGENEOUS}.
+     */
+    public static SubscriptionType subscriptionType(
+        Map<String, Integer> subscribedTopicNames
+    ) {
+        if (subscribedTopicNames.isEmpty()) {
+            return HOMOGENEOUS;
+        }
+
+        // Take the subscriber count of the first topic as the reference.
+        int referenceCount = subscribedTopicNames.values().iterator().next();
```

Comment:
Do we need to check if referenceCount is equal to member count?

Reply:
We don't allow subscribing with no topic names with the consumer protocol so this logic works in this case. However, thinking about https://github.com/apache/kafka/pull/15798, it may not be true when the classic protocol is used to join a consumer group. @dongnuo123 Do you confirm?

Reply:
> thinking about https://github.com/apache/kafka/pull/15798, it may not be true

Yes this is correct. I can add a check to prevent empty subscribed topic list form joining.

Reply:
For checking the referenceCount, is it possible to have
```
member A - topic-1, topic-2
member B - topic-2, topic-3
member C - topic-1, topic-3
```

Reply:
ah yes! I remember now that this is why I chose to do number of members

Reply:
we need to somehow get the accurate total member count and use this only

Reply:
> > thinking about #15798, it may not be true
> 
> Yes this isn't true. I can add a check to prevent empty subscribed topic list form joining.

This is not possible as we must support the current protocol as it is.

Reply:
> For checking the referenceCount, is it possible to have something like
> 
> ```
> member A - topic-1, topic-2
> member B - topic-2, topic-3
> member C - topic-1, topic-3
> ```

Good point. So we need to use the group size.